lib/Transforms/Scalar/DxilLoopUnroll.cpp

```diff
@@ -1163,13 +1163,79 @@ class DxilConditionalMem2Reg : public FunctionPass {
     return false;
   }
 
+  // Collect and remove all instructions that use AI, but
+  // give up if there are anything other than store, bitcast,
+  // or GEP.
+  static bool TryRemoveUnusedAlloca(AllocaInst *AI) {
+    std::vector<User *> WorkList;
+    std::vector<Instruction *> RemoveList;
+
+    for (User *U : AI->users()) {
+      WorkList.push_back(U);
+    }
+
+    while (WorkList.size()) {
+      User *U = WorkList.back();
+      WorkList.pop_back();
+      if (Instruction *UI = dyn_cast<Instruction>(U)) {
+        unsigned Opcode = UI->getOpcode();
+        if (Opcode == Instruction::BitCast ||
+          Opcode == Instruction::GetElementPtr ||
+          Opcode == Instruction::Store)
+        {
+          RemoveList.push_back(UI);
+          for (User *UU : U->users()) {
+            WorkList.push_back(UU);
+          }
+        }
+        else { // Load? PHINode? Memcpy? Assume written.
```

Comment:
`assume read`?  Also, we should be able to get rid of memcpy when ptr is dest as well, right?